File: src/main/java/io/github/przbetkier/tuscan/domain/match/MatchFullDetailsMapper.java
package io.github.przbetkier.tuscan.domain.match;

import io.github.przbetkier.tuscan.adapter.api.response.MatchFullDetailsResponse;
import io.github.przbetkier.tuscan.adapter.api.response.dto.MatchResult;
import io.github.przbetkier.tuscan.adapter.api.response.dto.Player;
import io.github.przbetkier.tuscan.adapter.api.response.dto.PlayerStats;
import io.github.przbetkier.tuscan.adapter.api.response.dto.Team;
import io.github.przbetkier.tuscan.adapter.api.response.dto.TeamStats;
import io.github.przbetkier.tuscan.client.match.MatchFullDetailsDto;
import io.github.przbetkier.tuscan.client.match.MatchStatsDto;
import io.github.przbetkier.tuscan.client.match.PlayerDto;

import java.math.BigDecimal;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import static io.github.przbetkier.tuscan.adapter.api.response.dto.MatchResult.LOSS;
import static io.github.przbetkier.tuscan.adapter.api.response.dto.MatchResult.WIN;
import static java.lang.Integer.valueOf;
import static java.util.Arrays.asList;

public class MatchFullDetailsMapper {

    private MatchFullDetailsMapper() {
    }

    public static MatchFullDetailsResponse map(MatchStatsDto matchStatsDto, String playerId) {
        MatchFullDetailsDto matchFullDetails = matchStatsDto.getMatchFullDetails().get(0);

        Team teamOne = getTeam(matchStatsDto, 0, getPlayers(matchStatsDto, 0));
        Team teamTwo = getTeam(matchStatsDto, 1, getPlayers(matchStatsDto, 1));

        return new MatchFullDetailsResponse(
                matchFullDetails.getMatchId(),
                matchFullDetails.getRoundStatsDto().getMap(),
                matchFullDetails.getRoundStatsDto().getScore(),
                valueOf(matchFullDetails.getRoundStatsDto().getRoundsCount()),
                asList(teamOne, teamTwo),
                matchFullDetails.getRoundStatsDto().getWinnerTeamId(),
                getResult(playerId, asList(teamOne, teamTwo), matchFullDetails.getRoundStatsDto().getWinnerTeamId()));
    }

    private static Player getPlayerFromPlayerDto(PlayerDto playerDto) {
        return new Player(
                playerDto.getPlayerId(), playerDto.getNickname(), new PlayerStats(
                valueOf(playerDto.getPlayerStats().getKills()),
                valueOf(playerDto.getPlayerStats().getAssists()),
                valueOf(playerDto.getPlayerStats().getDeaths()),
                valueOf(playerDto.getPlayerStats().getHeadshots()),
                valueOf(playerDto.getPlayerStats().getHeadshotPercentage()),
                mapKdRatio(playerDto),
                new BigDecimal(playerDto.getPlayerStats().getKrRatio()),
                valueOf(playerDto.getPlayerStats().getMvps()),
                valueOf(playerDto.getPlayerStats().getTripleKills()),
                valueOf(playerDto.getPlayerStats().getQuadroKills()),
                valueOf(playerDto.getPlayerStats().getPentaKills())));
    }

    private static BigDecimal mapKdRatio(PlayerDto playerDto) {
        String kdRatioString = playerDto.getPlayerStats().getKdRatio();
        BigDecimal kdRatio = new BigDecimal(kdRatioString);
        return kdRatio.compareTo(BigDecimal.ZERO) == 0
                ? new BigDecimal(playerDto.getPlayerStats().getKills())
                : kdRatio;
    }

    private static Team getTeam(MatchStatsDto matchStatsDto, Integer teamNumber, Set<Player> players) {
        return new Team(
                matchStatsDto.getMatchFullDetails().get(0).getTeams().get(teamNumber).getTeamId(),
                new TeamStats(
                        matchStatsDto.getMatchFullDetails().get(0).getTeams().get(teamNumber).getTeamStats().getTeamName(),
                        Double.valueOf(matchStatsDto.getMatchFullDetails().get(0).getTeams().get(teamNumber).getTeamStats().getHeadshotAvg())
                ),
                players);
    }

    private static Set<Player> getPlayers(MatchStatsDto matchStatsDto, Integer teamNumber) {
        return matchStatsDto.getMatchFullDetails().get(0).getTeams().get(teamNumber).getPlayers().stream().map(
                MatchFullDetailsMapper::getPlayerFromPlayerDto).collect(
                Collectors.toSet());
    }

    private static MatchResult getResult(String playerId, List<Team> teams, String winnerTeam) {
        Team winner = teams.stream().filter(t -> t.getTeamId().equals(winnerTeam)).findFirst().orElseThrow(RuntimeException::new);

        return winner.getPlayers().stream()
                .anyMatch(p -> p.getPlayerId().equals(playerId)) ? WIN : LOSS;
    }
}

File: src/main/java/io/github/przbetkier/tuscan/domain/match/MatchFullDetailsMapper.kt
package io.github.przbetkier.tuscan.domain.match

import io.github.przbetkier.tuscan.adapter.api.response.MatchFullDetailsResponse
import io.github.przbetkier.tuscan.adapter.api.response.dto.MatchResult
import io.github.przbetkier.tuscan.adapter.api.response.dto.Player
import io.github.przbetkier.tuscan.adapter.api.response.dto.PlayerStats
import io.github.przbetkier.tuscan.adapter.api.response.dto.Team
import io.github.przbetkier.tuscan.client.match.MatchDemoDto
import io.github.przbetkier.tuscan.client.match.MatchStatsDto
import io.github.przbetkier.tuscan.client.match.PlayerDto
import io.github.przbetkier.tuscan.domain.stats.DemoStats
import java.math.BigDecimal
import java.math.BigDecimal.ZERO

class MatchFullDetailsMapper {

    companion object {
        fun map(matchStatsDto: MatchStatsDto, playerId: String, matchDemoDto: MatchDemoDto): MatchFullDetailsResponse {
            val matchFullDetails = matchStatsDto.matchFullDetails.first()

            val teamOne = getTeam(matchStatsDto, 0, getPlayers(matchStatsDto, 0))
            val teamTwo = getTeam(matchStatsDto, 1, getPlayers(matchStatsDto, 1))

            return matchFullDetails.roundStatsDto.let {
                MatchFullDetailsResponse(
                        matchFullDetails.matchId,
                        it.map,
                        it.score,
                        it.roundsCount.toInt(),
                        listOf(teamOne, teamTwo),
                        it.winnerTeamId,
                        getResult(playerId, listOf(teamOne, teamTwo), it.winnerTeamId),
                        getDemo(matchDemoDto),
                        DemoStatus.NO_ACTION.name)
            }
        }

        private fun getDemo(matchDemoDto: MatchDemoDto?): String? {
            return matchDemoDto?.urls?.let { if (it.isNotEmpty()) it.first() else null }
        }

        private fun getPlayerFromPlayerDto(playerDto: PlayerDto): Player {
            return Player(
                    playerDto.playerId, playerDto.nickname,
                    playerDto.playerStats.let {
                        PlayerStats(
                                it.kills.toInt(),
                                it.assists.toInt(),
                                it.deaths.toInt(),
                                it.headshots.toInt(),
                                it.headshotPercentage.toInt(),
                                mapKdRatio(playerDto),
                                it.krRatio.toBigDecimal(),
                                it.mvps.toInt(),
                                it.tripleKills.toInt(),
                                it.quadroKills.toInt(),
                                it.pentaKills.toInt())
                    })
        }

        private fun mapKdRatio(playerDto: PlayerDto): BigDecimal {
            val kdRatio = playerDto.playerStats.kdRatio.toBigDecimal()
            return if (kdRatio.compareTo(ZERO) == 0) playerDto.playerStats.kills.toBigDecimal() else kdRatio
        }

        private fun getTeam(matchStatsDto: MatchStatsDto, teamNumber: Int, players: Set<Player>): Team {
            return matchStatsDto.matchFullDetails.first().teams[teamNumber].let {
                Team(it.teamId, it.teamStats.teamName, players)
            }
        }

        private fun getPlayers(matchStatsDto: MatchStatsDto, teamNumber: Int): Set<Player> {
            return matchStatsDto.matchFullDetails.first().teams[teamNumber].players.map {
                getPlayerFromPlayerDto(it)
            }.toSet()
        }

        private fun getResult(playerId: String, teams: List<Team>, winnerTeam: String): MatchResult {
            return if (teams.first { t -> t.teamId == winnerTeam }.players.any { p -> p.playerId == playerId })
                MatchResult.WIN else MatchResult.LOSS
        }
    }
}
